proxy/src/telemetry/metrics/mod.rs

```diff
@@ -597,14 +627,107 @@ impl HyperService for Serve {
                 .with_status(StatusCode::NotFound));
         }
 
-        let body = {
-            let metrics = self.metrics.lock()
-                .expect("metrics lock poisoned");
-            format!("{}", *metrics)
+        let metrics = self.metrics.lock()
+            .expect("metrics lock poisoned")
+            .deref()
+            .clone();
+
+        let is_gzip = is_gzip(&req);
+        let (writer, body) = WriteBody::new();
+        let work = future::lazy(move || {
+            if is_gzip {
+                let mut writer = GzEncoder::new(Vec::<u8>::new(), Compression::Default);
```

Comment:
What is the difference in compression between the default compression level and `CompressionOptions::fast()`? I think unless there is a huge difference, we should prefer the lower-CPU-usage variant. Because of the nature of the data we're compressing and the way the deflate algorithm works, I expect that even the "fast" level will significantly shrink the output.